requirements.txt: Update Errbot

nvzard opened the pull request:
Closes https://github.com/coala/corobo/issues/571
Closes https://github.com/coala/corobo/issues/584

nvzard commented:
https://github.com/errbotio/errbot/commit/9a03b72a33c12a34c037c0f4286d5192c3f766b6#diff-86fb3ee3298c4b1416b1da174131201c 
This removed the `instanciateElement` method from the plugin manager.

nvzard commented on tests/helper.py:
```diff
@@ -1,8 +1,8 @@
     testbot = TestBot(loglevel=loglevel, extra_config=config)
     testbot.start()
-    plug = testbot.bot.plugin_manager.instanciateElement(klass)
+    plug = klass(testbot.bot, plugin_name)
```
```
def instanciateElement(self, element) -> BotPlugin:
    """Overrides the instanciation of plugins to inject the bot reference."""
    return element(self.bot, name=self._current_pluginfo.name)
```

nvzard replied:
@meetmangukiya @gbin does it make a difference, where a class gets instantiated?

nvzard replied:
No this is not the cause for failures. Facing the `TemplateNotFound` exception due to the changes made to `errbot.templating` by this commit https://github.com/errbotio/errbot/commit/9a03b72a33c12a34c037c0f4286d5192c3f766b6#diff-b27e6e3f6d8b53f810de95de950feea9

meetmangukiya replied:
Figured a workaround, explore yourself further:

```
From d9858f2dc7162273d8315ae37bfee6467257599e Mon Sep 17 00:00:00 2001
From: Meet Mangukiya <meetmangukiya98@gmail.com>
Date: Wed, 11 Jul 2018 06:44:23 +0530
Subject: [PATCH] Fix templating issue

---
 tests/labhub_test.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/tests/labhub_test.py b/tests/labhub_test.py
index 9d69a64..e06b195 100644
--- a/tests/labhub_test.py
+++ b/tests/labhub_test.py
@@ -19,9 +19,21 @@ from plugins.labhub import LabHub
 
 from tests.helper import plugin_testbot
 
+
+import errbot
+from errbot.plugin_info import PluginInfo
+from errbot.templating import add_plugin_templates_path
+from pathlib import Path
+
+
 class TestLabHub(unittest.TestCase):
 
     def setUp(self):
+        labhub_info_path = Path(__file__).parent / '..' / 'plugins' / 'labhub.plug'
+        with labhub_info_path.open() as plugfile:
+            labhub_info = PluginInfo.load_file(plugfile, labhub_info_path)
+        add_plugin_templates_path(labhub_info)
+
         plugins.labhub.github3 = create_autospec(github3)
 
         self.mock_org = create_autospec(github3.orgs.Organization)
-- 
2.18.0
```

jayvdb commented on tests/labhub_test.py:
```diff
@@ -22,6 +26,12 @@
 class TestLabHub(unittest.TestCase):
 
     def setUp(self):
+        labhub_info_path = (Path(__file__)
```
This should be put into a helper somewhere.

nvzard replied:
I shifted this to `CoroboTestHelper` class in the commit I am about to push once https://github.com/coala/corobo/pull/573/ gets merged. But I'll move this into `helpers.py` for now.

meetmangukiya replied:
Doesn't make much sense because this is not generic and needed in all plugins using the helper. Specific to plugin stays specific to test cases, no?

meetmangukiya commented on tests/helper.py:
```diff
@@ -1,8 +1,18 @@
+from pathlib import Path
 
-def plugin_testbot(klass, loglevel, config=None):
+def plugin_testbot(klass, plugin_name, loglevel, config=None):
```
can't the plugin name be derived using the `klass`?

nvzard replied:
but sometimes the plugin name can be different from the class name. Eg: git_stats plugin

meetmangukiya replied:
There is a solution, load plugin file, get the name key.

nvzard replied:
but even for that, we'll have to pass the name of the plug file.

jayvdb replied:
Can you not find the name of the plug file from the plug name?

nvzard replied:
Yes, for labhub. But it is not required to load templates for  Ban and git_stats plugins.

nvzard commented:
I simply rebased my https://github.com/coala/corobo/pull/581 commit over this PR and resolved the merge conflict. Should I also delete the changes made to lines in `requirements.txt: Update Errbot` which are again being changed by the top commit?

meetmangukiya commented on tests/corobo_test_case.py:
```diff
@@ -0,0 +1,36 @@
+
+    def setUp(self, plug_filename):
+        super().setUp(loglevel=logging.ERROR,
+                      extra_config={'BACKEND': 'text'})
```
text backend? why?

nvzard replied:
Because `test_create_issie_cmd` needed text backend https://github.com/coala/corobo/blob/d697e18d436a33dca07f29e6d26544d3234750f3/tests/labhub_test.py#L165.

If I had not used it. I would have to call `super().setup(...)`at the beginning of every test case.

meetmangukiya commented on tests/corobo_test_case.py:
```diff
@@ -0,0 +1,36 @@
+
+class CoroboTestCase(FullStackTest):
+
+    def setUp(self, plug_filename):
```
Replace the parameter with plugin class

meetmangukiya replied:
Or a `plugins` set or a list, an iterable, for loading of plugins.

nvzard replied:
I do not understand this one clearly. What is the purpose of this since we need only one plugin class at once?
>Or a plugins set or a list, an iterable, for loading of plugins.

meetmangukiya replied:
If we are testing dependent plugins? We need to mock more than one plugins, no?

nvzard replied:
Yes right. Got it :+1:

meetmangukiya commented on tests/corobo_test_case.py:
```diff
@@ -0,0 +1,36 @@
+                      extra_config={'BACKEND': 'text'})
+        self.load_plug_file(plug_filename)
+
+    def load_plug_file(self, plug_filename):
```
Since we are accepting plugin class, here's a way you can determine the filename from that. 
```py
filename = klass.__module__.split('.')[-1] + '.plug'
```

meetmangukiya replied:
Rename to `load_plugin_templates`

meetmangukiya commented on tests/git_stats_test.py:
```diff
@@ -14,12 +12,13 @@
+class TestGitStats(CoroboTestCase):
 
     def setUp(self):
+        super().setUp(plug_filename='git_stats.plug')
```
Adapt the changes I suggested to the `CoroboTestCase` class

meetmangukiya commented on tests/helper.py:
```diff
@@ -1,9 +1,21 @@
+from pathlib import Path
 
 
 def plugin_testbot(klass, loglevel, config=None):
```
Do we even need this after `CoroboTestCase`?

nvzard replied:
Nope

meetmangukiya commented on tests/helper.py:
```diff
@@ -1,9 +1,21 @@
     return plug, testbot
+
+
+def load_templates(plug_file_name):
```
Same, do we even need this after `CoroboTestCase`?

nvzard replied:
Nope

meetmangukiya commented on tests/labhub_test.py:
```diff
@@ -47,62 +53,67 @@ def test_invite_cmd(self):
+            'TEAMS': teams,
+            'is_room_member': MagicMock(),
+        }
+        labhub = self.load_plugin(
```
load plugin only once, not in each test case. There is no `tearDown` unloading the plugin. Load the plugin in `setUp` with global mocks that are required everywhere. and then directly use `self.inject_mocks` for subsequent mock injection.

Applies for everywhere.

jayvdb commented on .moban.yaml:
```diff
@@ -4,7 +4,7 @@ name: corobo
 
 dependencies:
   - pyopenssl<17.5
```
please investigate why we had this pin.  and try to remove it.
may be it isnt needed any more

nvzard replied:
okay, I'll look into it.

nvzard replied:
https://github.com/coala/corobo/pull/592

jayvdb replied:
Remove it entirely in this PR

jayvdb commented:
Do we need all of https://github.com/coala/corobo/pull/572/commits/56995e1c28dde9060551a4bd27614086578a793d in this PR?

errbot upgrade fails without it?

nvzard commented:
@jayvdb I rebased the commit in https://github.com/coala/corobo/pull/581 on top of this PR as you asked https://github.com/coala/corobo/pull/581#pullrequestreview-137563815

https://github.com/coala/corobo/commit/56995e1c28dde9060551a4bd27614086578a793d is the commit from the PR https://github.com/coala/corobo/pull/581 including the requested changes.

I'll have to close https://github.com/coala/corobo/pull/581 since I rebased that commit on top of this PR.

jayvdb commented on tests/labhub_test.py:
```diff
@@ -50,69 +65,71 @@ def test_invite_cmd(self):
-        testbot.assertCommand('!invite meet to newcomers',
-                              '@meet is not a member of this room.')
+        mock_dict['is_room_member'].return_value = False
+        self.assertCommand('!invite meet to newcomers',
```
add `testbot = self` above these, so that you dont need to edit so many lines

jayvdb commented on tests/labhub_test.py:
```diff
@@ -161,56 +157,50 @@ def test_create_issue_cmd(self):
+        plugins.labhub.GitHubToken.assert_called_with(None)
+        plugins.labhub.GitLabPrivateToken.assert_called_with(None)
+
+        self.assertCommand(
```
`testbot_public = self`

jayvdb commented on tests/labhub_test.py:
```diff
@@ -227,97 +217,88 @@ def test_is_newcomer_issue(self):
         mock_iss.unassign = MagicMock()
         self.mock_team.is_member.return_value = True
 
-        testbot.assertCommand(
```
`testbot = self`

jayvdb commented on tests/ban_test.py:
```diff
@@ -27,18 +27,13 @@ def test_ban_cmd(self, mockjson, mockreq):
             {'id': '897', 'uri': 'coala/coala-bears'}
         ]
         mockjson.loads.return_value = fake_room_data
-        testbot.assertCommand('!ban @nvzard',
```
`testbot = self`

jayvdb commented on tests/ban_test.py:
```diff
@@ -49,6 +44,6 @@ def test_unban_cmd(self, mockjson, mockreq):
             {'id': '897', 'uri': 'coala/coala-bears'}
         ]
         mockjson.loads.return_value = fake_room_data
-        testbot.assertCommand('!unban @nvzard',
```
`testbot = self` (and similar throughout) will make this patch much smaller, easier to review, allowing reviewers to focus on the more important other changes that are occurring.

jayvdb commented on tests/git_stats_test.py:
```diff
@@ -53,10 +50,10 @@ def test_pr_list(self):
         self.mock_repo.merge_requests = [mock_github_mr]
 
         # Non-existing repo
-        testbot.assertCommand(cmd_github.format('b'),
```
`testbot = self`

jayvdb commented:
ack https://github.com/coala/corobo/pull/572/commits/a9d99def19266d2982661e2fbafa2dee3bcc9152

jayvdb commented:
@gitmate-bot ff

nvzard commented on tests/labhub_test.py:
```diff
@@ -161,14 +162,8 @@ def test_create_issue_cmd(self):
         # Stop ignoring
```
@jayvdb ACL's are working as they are supposed to. But it is not actually possible to test ACLs since the message/ACL cannot be mocked from here.
ACL is well tested by the framework as it is an upstream plugin and tested upstream.
